txpipe/cosmos_weight.py

```diff
@@ -0,0 +1,232 @@
+                      'n_neighbors': 10}
+
+    def run(self):
+        """
+        Main function.
+        This stage matches the COSMOS-30band data with the HSC COSMOS sample
+        cut with the same criteria as our data and produces colour-space
+        weights to match our sample so it can be used to estimate redshift
+        distributions.
+        """
+
+        # Read HSC COSMOS catalog
+        logger.info("Reading HSC COSMOS catalog.")
+        cat_photo = Table.read(self.get_input('cosmos_hsc'))
+        # Clean nulls and nans
+        sel = np.ones(len(cat_photo), dtype=bool)
+        isnull_names = []
+        for key in cat_photo.keys():
+            if key.__contains__('isnull'):
+                sel[cat_photo[key]] = 0
+                isnull_names.append(key)
+            else:
+                # Keep photo-z's even if they're NaNs
+                if not key.startswith("pz_"):
+                    sel[np.isnan(cat_photo[key])] = 0
+        cat_photo.remove_columns(isnull_names)
+        cat_photo.remove_rows(~sel)
+        # Select lens objects
+        lens_cat = self.select_lens(cat_photo)
+        cat_photo = cat_photo[lens_cat]
+
+        # Define shear catalog
+        cat_photo['shear_cat'] = self.shear_cut(cat_photo)
+
+        ####
+        # Read COSMOS-30band
+        logger.info("Reading COSMOS 30band catalog.")
+        cat30 = fits.open(self.get_input('cosmos_data'))[1].data
+        lim_indices = np.where((0.01 < cat30['PHOTOZ']) &
+                               (9 > cat30['PHOTOZ']) &
+                               (cat30['TYPE'] == 0) &
+                               (cat30['ZP_2'] < 0) &
+                               (cat30['MASS_BEST'] > 7.5) &
+                               (np.maximum(cat30['ZPDF_H68']-cat30['ZPDF'],
+                                           cat30['ZPDF']-cat30['ZPDF_L68']) <
+                                0.05*(1+cat30['PHOTOZ'])) &
+                               (cat30['CHI2_BEST'] < cat30['CHIS']) &
+                               (cat30['CHI2_BEST']/cat30['NBFILT'] < 5.))
+        cat30 = cat30[lim_indices]
+
+        ####
+        # Match coordinates
+        logger.info("Matching coordinates.")
+        cosmos_crd = SkyCoord(ra=np.array(cat30['ALPHA_J2000'])*u.deg,
+                              dec=np.array(cat30['DELTA_J2000'])*u.deg)
+        hsc_crd = SkyCoord(ra=np.array(cat_photo['ra'])*u.deg,
+                           dec=np.array(cat_photo['dec'])*u.deg)
+        # Nearest neighbors
+        cosmos_index, dist_2d, dist_3d = hsc_crd.match_to_catalog_sky(cosmos_crd)
+        # Cut everything further than 1 arcsec
+        mask = dist_2d.degree*60*60 < 1
+        cat_photo_good = cat_photo[mask]
+        cat30_good = cat30[cosmos_index[mask]]
+        cosmos_photo_index_matched = cosmos_index[mask]
+
+        t1 = Table.from_pandas(pd.DataFrame(cat30_good))
+        keys_t2 = ['gcmodel_mag', 'rcmodel_mag', 'icmodel_mag', 'zcmodel_mag',
```

Comment:
Maybe keys_t2 can be part of the config options? I don't think that we will use it with a different photo-z algorithm but it can be possible. I don't know. What do you think @anicola?